src/main/drivers/serial_uart_stm32f30x.c

```diff
@@ -212,322 +332,124 @@ void serialUARTInit(IO_t tx, IO_t rx, portMode_t mode, portOptions_t options, ui
-    s->rxDMAChannel = DMA1_Channel5;
-    s->rxDMAPeripheralBaseAddr = (uint32_t)&s->USARTx->RDR;
+#if defined(USE_UART1_RX_DMA) || defined(USE_UART2_RX_DMA) || defined(USE_UART3_RX_DMA)
+    // This doesn't work... Was original RX DMA working at all???
+    if (uartDev->rxDMAChannel) {
+        s->rxDMAChannel = uartDev->rxDMAChannel;
+        s->rxDMAPeripheralBaseAddr = (uint32_t)&s->USARTx->RDR;
+        dmaInit(dmaGetIdentifier(uartDev->rxDMAChannel), OWNER_SERIAL_RX, RESOURCE_INDEX(device));
+    }
 #endif
-#ifdef USE_UART1_TX_DMA
-    s->txDMAChannel = DMA1_Channel4;
-    s->txDMAPeripheralBaseAddr = (uint32_t)&s->USARTx->TDR;
+
+// Get rid of this when DMA become configurable
+#if defined(USE_UART1_TX_DMA) || defined(USE_UART2_TX_DMA) || defined(USE_UART3_TX_DMA)
+    if (uartDev->txDMAChannel) {
+        s->txDMAChannel = uartDev->txDMAChannel;
+        s->txDMAPeripheralBaseAddr = (uint32_t)&s->USARTx->TDR;
+        const dmaIdentifier_e identifier = dmaGetIdentifier(uartDev->txDMAChannel);
+
+        dmaInit(identifier, OWNER_SERIAL_TX, RESOURCE_INDEX(device));
+        dmaSetHandler(identifier, handleUsartTxDma, uartDev->txPriority, (uint32_t)s);
+    }
 #endif
 
-    RCC_ClockCmd(RCC_APB2(USART1), ENABLE);
+    RCC_ClockCmd(uartDev->rcc, ENABLE);
 
 #if defined(USE_UART1_TX_DMA) || defined(USE_UART1_RX_DMA)
+// Is this handled in dmaInit or dmaSetHandler?
     RCC_ClockCmd(RCC_AHB(DMA1), ENABLE);
 #endif
 
-    serialUARTInit(IOGetByTag(IO_TAG(UART1_TX_PIN)), IOGetByTag(IO_TAG(UART1_RX_PIN)), mode, options, GPIO_AF_7, 1);
-
-#ifdef USE_UART1_TX_DMA
-    dmaInit(DMA1_CH4_HANDLER, OWNER_SERIAL, 1);
-    dmaSetHandler(DMA1_CH4_HANDLER, handleUsartTxDma, NVIC_PRIO_SERIALUART1_TXDMA, (uint32_t)&uartPort1);
-#endif
+    serialUARTInitIO(IOGetByTag(uartDev->tx), IOGetByTag(uartDev->rx), mode, options, uartDev->af, device);
 
-#ifndef USE_UART1_RX_DMA
-    NVIC_InitTypeDef NVIC_InitStructure;
+    if (!uartDev->rxDMAChannel) {
+        NVIC_InitTypeDef NVIC_InitStructure;
 
-    NVIC_InitStructure.NVIC_IRQChannel = USART1_IRQn;
-    NVIC_InitStructure.NVIC_IRQChannelPreemptionPriority = NVIC_PRIORITY_BASE(NVIC_PRIO_SERIALUART1_RXDMA);
-    NVIC_InitStructure.NVIC_IRQChannelSubPriority = NVIC_PRIORITY_SUB(NVIC_PRIO_SERIALUART1_RXDMA);
-    NVIC_InitStructure.NVIC_IRQChannelCmd = ENABLE;
-    NVIC_Init(&NVIC_InitStructure);
-#endif
+        NVIC_InitStructure.NVIC_IRQChannel = uartDev->irqn;
+        NVIC_InitStructure.NVIC_IRQChannelPreemptionPriority = NVIC_PRIORITY_BASE(uartDev->rxPriority);
+        NVIC_InitStructure.NVIC_IRQChannelSubPriority = NVIC_PRIORITY_SUB(uartDev->rxPriority);
+        NVIC_InitStructure.NVIC_IRQChannelCmd = ENABLE;
+        NVIC_Init(&NVIC_InitStructure);
+    }
 
     return s;
 }
-#endif
 
-#ifdef USE_UART2
-uartPort_t *serialUART2(uint32_t baudRate, portMode_t mode, portOptions_t options)
+#ifdef USE_UART1
+uartPort_t *serialUART1(uint32_t baudRate, portMode_t mode, portOptions_t options)
```

Comment:
I think I'll do it when F1 and F7 are ready. I want to clean up the terrible reference to `USART_TypeDef` for `uartOpen` at the same time, too. Thoughts?